### PR TITLE
Prototype `buildSrc/groovy/dart/*` scripts in Kotlin

### PR DESCRIPTION
**Pull request by yevhenii-nadtochii**
This draft is not going to be merged as approved. A separate PR into config repository would be created.

### CODE REVIEW

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/Dart.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+fun Project.dart(configuration: DartExtension.() -> Unit) {
+    extensions.run {
+        configuration.invoke(
+            findByType() ?: create("dartExtension", project)
+        )
+    }
+}
+
+/**
+ * Encapsulates Dart related configuration.
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/Dart.kt]: Let's not use this "encapsulates" at all. It tells nothing. I would start like this: 

``` 
Gradle extension which ...
```

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/Dart.kt [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+package io.spine.internal.gradle.dart
+
+import org.apache.tools.ant.taskdefs.condition.Os
+import org.gradle.api.Project
+import org.gradle.api.tasks.TaskContainer
+import org.gradle.kotlin.dsl.create
+import org.gradle.kotlin.dsl.findByType
+
+/**
+ * Defines the scope for Dart related configuration.
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/Dart.kt]: It's "Dart-related". Here and in other documentation pieces.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/task/Build.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ * List of tasks to be created:
+ *
+ *  1. `resolveDependencies` - fetches the dependencies declared via `pubspec.yaml`;
+ *  2. `cleanPackageIndex` - deletes the resolved `.packages` file on this Dart module;
+ *  3. `testDart` - runs Dart tests declared in the `./test` directory.
+ *
+ * Usage example:
+ *
+ * ```
+ * dart {
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/task/Build.kt:57]: Do we need users to import anything? If so, please specify.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/task/Publish.kt [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+        from(project.projectDir) {
+            include("**/*.dart", "pubspec.yaml", "**/*.md")
+            exclude("proto/", "generated/", "build/", "**/.*")
+        }
+        from("${project.rootDir}/LICENSE")
+        into(publicationDirectory)
+
+        doLast {
+            logger.debug("Prepared Pub publication in directory `$publicationDirectory`.")
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/task/Publish.kt]:
```
Pub publication is prepared in the directory `...`.
```

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/task/Publish.kt [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+private fun DartTasks.stagePubPublication(): Task =
+    create<Copy>("stagePubPublication") {
+        description = "Prepares the Dart package for Pub publication."
+        group = dartPublishTask
+
+        from(project.projectDir) {
+            include("**/*.dart", "pubspec.yaml", "**/*.md")
+            exclude("proto/", "generated/", "build/", "**/.*")
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/task/Publish.kt:107]: I would document this somewhere. It seems important.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/task/Publish.kt [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        group = dartPublishTask
+
+        workingDir(publicationDirectory)
+        commandLine(pubExecutable, "publish", "--trace")
+
+        val sayYes = "y".byteInputStream()
+        standardInput = sayYes
+    }
+
+private fun DartTasks.activateLocally(): Task =
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/task/Publish.kt]: Please document this one. It's not very obvious. I would expect a link to the description of local activation — if that's a thing.

File: client/build.gradle.kts [hunk 7]
```diff
@@ -60,6 +61,13 @@ dependencies {
     testProtobuf("io.spine.tools:spine-tool-base:$spineBaseVersion")
 }
 
+dart {
+    tasks {
+        registerBuildTasks()
```
Comment (armiol) [inline client/build.gradle.kts]: Can we turn it into this?
```
register(build(), publish())
```
It's two "register" (and I want to have just one), and two redundant "Tasks" (which is the context in which we operate already) to kill.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/Dart.kt [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    val packageIndex: String
+}
+
+/**
+ * Context for assembling Dart related tasks with access to the current [DartEnvironment].
+ */
+class DartTasks(dartEnv: DartEnvironment, tasks: TaskContainer)
+    : DartEnvironment by dartEnv, TaskContainer by tasks
+{
+    val dartBuildTask = "Dart/Build"
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/Dart.kt]: I would consider reducing the visibility of the constants to `internal`.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/Dart.kt [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+open class DartExtension(project: Project) {
+
+    private val defaultEnv = object : DartEnvironment {
+        override val pubExecutable = "pub${if (Os.isFamily(Os.FAMILY_WINDOWS)) ".bat" else ""}"
+        override val packageIndex = "${project.projectDir}/.packages"
+        override val pubSpec = "${project.projectDir}/pubspec.yaml"
+        override val publicationDirectory = "${project.buildDir}/pub/publication/${project.name}"
+    }
+
+    @Suppress("LeakingThis")
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/Dart.kt]: Please comment on suppressions like this one.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/Dart.kt [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    private val tasks = DartTasks(defaultEnv, project.tasks)
+
+    fun tasks(configuration: DartTasks.() -> Unit) = configuration.invoke(tasks)
+}
+
+/**
+ * Information about Dart environment.
+ */
+interface DartEnvironment {
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/dart/Dart.kt]: We need an example on how to customize it.

**Review by armiol — CHANGES_REQUESTED** [review]
@yevhenii-nadtochii LGTM in general. Please see my comments. 

I'd like to review this draft one more time once the comments are addressed.

**Comment by yevhenii-nadtochii** [comment]
Closing this draft as an associated PR into `config` repository is created.
See [Migrate `buildSrc/.../groovy/dart/*` scripts to Kotlin ](https://github.com/SpineEventEngine/config/pull/311)

**Pull request closed without merging**